internal/function_net.go

```
@@ -0,0 +1,156 @@
+	}
+	if length < 0 || int(length) > len(v)*8 {
+		return nil, fmt.Errorf("NET.IP_TRUNC: length must be in the range from 0 to %d", len(v)*8)
+	}
+	ip := net.IP(v)
+	mask := net.CIDRMask(int(length), len(v)*8)
+	return BytesValue(ip.Mask(mask)), nil
+}
+
+func NET_IPV4_FROM_INT64(v int64) (Value, error) {
+	ip := make([]byte, 4)
+	binary.BigEndian.PutUint32(ip, uint32(v))
+	return BytesValue(ip), nil
+}
+
+func NET_IPV4_TO_INT64(v []byte) (Value, error) {
+	if len(v) != 4 {
+		return nil, fmt.Errorf("NET.IPV4_TO_INT64: length of bytes array must be 4")
+	}
+	return IntValue(binary.BigEndian.Uint32(v)), nil
+}
+
+func NET_PUBLIC_SUFFIX(v string) (Value, error) {
+	parsed := parse_url(v)
+	if parsed == nil {
+		return nil, nil
+	}
+	host := parsed.Hostname()
+	suffix := public_suffix(host)
+	if suffix == "" {
+		return nil, nil
+	}
+	return StringValue(suffix), nil
+}
+
+func NET_REG_DOMAIN(v string) (Value, error) {
+	parsed := parse_url(v)
+	if parsed == nil {
+		return nil, nil
+	}
+	host := parsed.Hostname()
+	split_host := strings.Split(host, ".")
+	suffix := public_suffix(host)
+	split_suffix := strings.Split(suffix, ".")
+	if host == "" || suffix == "" || len(split_host) <= len(split_suffix) {
+		return nil, nil
+	}
+	return StringValue(strings.Join(split_host[len(split_host)-len(split_suffix)-1:], ".")), nil
+}
+
+func NET_SAFE_IP_FROM_STRING(v string) (Value, error) {
+	ip := parse_ip(v)
+	if ip == nil {
+		return nil, nil
+	}
+	return BytesValue(ip), nil
+}
+
+func parse_url(v string) *url.URL {
+	parsed, err := url.Parse(v)
+	if err != nil {
+		return nil
+	}
+	if parsed.Host == "" {
+		parsed, err = url.Parse("//" + strings.TrimSpace(v))
+		if err != nil {
+			return nil
+		}
+	}
+	return parsed
+}
+
+func public_suffix(host string) string {
+	reg, _ := regexp.Compile(`[^.]\.{2,}[^.]`)
+	if reg.MatchString(host) {
+		return ""
+	}
+	split_host := strings.Split(host, ".")
+	encoded, err := idna.ToASCII(strings.ToLower(host))
+	if err != nil {
+		return ""
```

Is it ok to ignore this error?

---

Fixed in https://github.com/goccy/go-zetasqlite/pull/99/commits/b68098425acba41999f40c821ed0e588556dbdef

The logic is clearer when an error is returned 😃 

In addition, the document doesn't specify that `NET.PUBLIC_SUFFIX` returns NULL when normalization fails.
So, the error shouldn't be suppressed I guess.